### DISCUSSION TITLE
How does the start-time metric work within the context of react-native-firebase

### DISCUSSION THREAD

**Discussion by dannyhw** [invertase/react-native-firebase#5483] [Q&A] [answered]
I apologise if this is overly generic but I'm not sure whether there is specific handling of this in the perf package or if it's just like a wrapper around the firebase sdks. I've been searching through a lot of documentation and it's still unclear to me how this works.

What I'm wondering is if there is something that hooks into the react native lifecycle to get an accurate start time reading, for example only after the first render completes. Or is it more like default readings from the underlying sdk?

From what I can tell it relies on the notification UIApplicationDidBecomeActiveNotification, however I'm not sure where this gets triggered and if there is something that triggers that in this package or if it's automatic for all apps or by react-native.

Thanks for your time and again if this question is out of place here I'm sorry 🙇  .

**Comment by mikehardy** [comment, marked as answer]
You might like: https://github.com/invertase/react-native-firebase/discussions/5425

**Reply by dannyhw** [reply]
I had already seen that but the question was worded really weirdly so figured it wasn't the same. So based on your answer there then there is nothing specific in this library to the start-up time. Honestly still not sure how the startup metric works and the firebase docs don't help much. I suppose I'll do some experiments and figure out from there.

Thanks anyway.

**Reply by mikehardy** [reply]
Experiments are the only way yes, but I am really curious for the results - just they should be posted on that discussion I think, to centralize.

I am going to *guess* (read as: verify with an experiment) that performance library will stop the built-in / default app start timer as soon as MainActivity loads, which from my point of view is an incomplete picture in a react-native context as that only just begins the JS bundle load and react-native View render.

I am going to *guess* (again, test needed) that either adding a custom timer start in AppDelegate.m and MainApplication.java and a custom timer finish on your App.js - called whenever you know the first screen is fully rendered - will be the way to get a single timer full app start timing.

Alternatively / testing at same time for comparison - putting a timer start in index.js and similar in App.js (or however your app is structured) such that it stops after full render of initial page may be sufficient in combination with the default app start perf timer that the module gives you for free, if two timers added together is sufficient / and/or the granularity of seeing JS timing vs Native timing is useful. In this case there may be a gap though between whenever the native timer stops and the JS one starts, if my guesses are correct on how it's going to work.

But no one has actually done the tests, and the results will be really interesting I think

**Reply by dannyhw** [reply]
Yeah I've been doing some research and my current feeling is that it is at the point of main activity being loaded.
I found this suggested in a blog post
https://github.com/bamlab/react-native-performance/tree/master/react-native-startup-trace

They seem to do some kind of custom trace, and I'm still investigating the code etc. It looks like an interesting approach, so I'll probably try it out soon and see how it behaves. 

I haven't done the testing yet, however I'm currently in the investigation/documentation stage of this journey 😅 .
Once I have some results I'll add my findings on the other discussion.

Thanks for your time 🙏
